libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/JsonExtensions.cs

```diff
@@ -36,31 +36,48 @@ public static T Dequeue<T>(this List<T> queue)
         /// Replaces the binding paths in a JSON Token value with the evaluated results recursively. Returns the final JSON Token value.
         /// </summary>
         /// <param name="token">A JSON Token value which may have some binding paths.</param>
+        /// <param name="valueExpr">Original ValueExpression.</param>
         /// <param name="state">A scope for looking up variables.</param>
         /// <returns>Deep data binding result.</returns>
-        public static JToken ReplaceJTokenRecursively(this JToken token, object state)
+        public static JToken ReplaceJTokenRecursively(this JToken token, ValueExpression valueExpr, object state)
+        {
```

Comment:
The result seems to be right, but i'm not sure we need to expand this interface or we have to introduce a new property "OriginalValue". 

Can we we simply guard on the `token`'s type here?